File: backend/app/models/work_schedule.py
"""Work schedule models for project timeline control."""
from app.core.timeutil import utc_now
import enum
import uuid
from datetime import date, datetime

from sqlalchemy import Boolean, Date, DateTime, Enum, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


def _uuid() -> str:
    return str(uuid.uuid4())


class WorkScheduleStatus(str, enum.Enum):
    draft = "draft"
    submitted = "submitted"
    confirmed = "confirmed"
    rejected = "rejected"
    archived = "archived"


class WorkScheduleItemStatus(str, enum.Enum):
    planned = "planned"
    ready = "ready"
    in_progress = "in_progress"
    submitted = "submitted"
    accepted = "accepted"
    delayed = "delayed"
    blocked = "blocked"
    cancelled = "cancelled"


class ProjectWorkSchedule(Base):
    __tablename__ = "project_work_schedules"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("projects.id"), index=True, nullable=False)
    status: Mapped[WorkScheduleStatus] = mapped_column(
        Enum(WorkScheduleStatus), default=WorkScheduleStatus.draft, index=True, nullable=False
    )
    title: Mapped[str] = mapped_column(String(255), default="План-график работ")
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    planned_start_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    planned_finish_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    rejection_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_by: Mapped[str] = mapped_column(String(36), ForeignKey("users.id"), index=True)
    submitted_by: Mapped[str | None] = mapped_column(String(36), ForeignKey("users.id"), nullable=True)
    confirmed_by: Mapped[str | None] = mapped_column(String(36), ForeignKey("users.id"), nullable=True)
    rejected_by: Mapped[str | None] = mapped_column(String(36), ForeignKey("users.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utc_now)
    submitted_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    confirmed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    rejected_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=utc_now)
    # E5: monotonic version; supersedes_id → previous schedule row when re-issued after reject
    schedule_version: Mapped[int] = mapped_column(Integer, default=1)
    supersedes_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("project_work_schedules.id"), nullable=True, index=True
    )


class ProjectWorkScheduleItem(Base):
    __tablename__ = "project_work_schedule_items"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    schedule_id: Mapped[str] = mapped_column(String(36), ForeignKey("project_work_schedules.id"), index=True, nullable=False)
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("projects.id"), index=True, nullable=False)
    stage_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("stages.id"), nullable=True, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[WorkScheduleItemStatus] = mapped_column(
        Enum(WorkScheduleItemStatus), default=WorkScheduleItemStatus.planned, index=True, nullable=False
    )
    planned_start_date: Mapped[date] = mapped_column(Date, nullable=False)
    planned_finish_date: Mapped[date] = mapped_column(Date, nullable=False)
    actual_start_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    actual_finish_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    depends_on_item_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("project_work_schedule_items.id"), nullable=True, index=True
    )
    requires_customer_acceptance: Mapped[bool] = mapped_column(Boolean, default=True)
    requires_photo: Mapped[bool] = mapped_column(Boolean, default=True)
    requires_hidden_work_acceptance: Mapped[bool] = mapped_column(Boolean, default=False)
    delay_days: Mapped[int] = mapped_column(Integer, default=0)
    blocking_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    progress_percent: Mapped[float] = mapped_column(Float, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=utc_now)

File: apps/mobile/lib/api/workSchedule.ts
import { req, ApiError } from './client';
import { shouldQueueReplaySafeMutation } from './failurePolicy';
import { createClientRequestId } from '@/lib/clientRequestId';

export type WorkScheduleStatus = 'draft' | 'submitted' | 'confirmed' | 'rejected' | 'archived';
export type WorkScheduleItemStatus =
  | 'planned'
  | 'ready'
  | 'in_progress'
  | 'submitted'
  | 'accepted'
  | 'delayed'
  | 'blocked'
  | 'cancelled';

export type WorkScheduleItem = {
  id: string;
  schedule_id: string;
  project_id: string;
  stage_id?: string | null;
  title: string;
  description?: string | null;
  status: WorkScheduleItemStatus;
  planned_start_date: string;
  planned_finish_date: string;
  actual_start_date?: string | null;
  actual_finish_date?: string | null;
  depends_on_item_id?: string | null;
  requires_customer_acceptance: boolean;
  requires_photo: boolean;
  requires_hidden_work_acceptance: boolean;
  delay_days: number;
  blocking_reason?: string | null;
  sort_order: number;
  progress_percent: number;
  created_at: string;
  updated_at: string;
};

export type WorkSchedule = {
  id: string;
  project_id: string;
  status: WorkScheduleStatus;
  title: string;
  description?: string | null;
  planned_start_date?: string | null;
  planned_finish_date?: string | null;
  rejection_reason?: string | null;
  created_by: string;
  submitted_by?: string | null;
  confirmed_by?: string | null;
  rejected_by?: string | null;
  created_at: string;
  submitted_at?: string | null;
  confirmed_at?: string | null;
  rejected_at?: string | null;
  updated_at: string;
  items: WorkScheduleItem[];
};

export type ActiveWorkScheduleResult =
  | { kind: 'absent' }
  | { kind: 'plan'; plan: WorkSchedule };

export type WorkScheduleCreateInput = Partial<WorkSchedule> & { client_request_id?: string };

async function queueReplaySafeScheduleCommand<T>(
  userId: string,
  path: string,
  body: string,
  error: unknown,
): Promise<T> {
  if (!shouldQueueReplaySafeMutation(error)) throw error;
  const { enqueue } = await import('@/lib/offlineQueue');
  await enqueue({ path, method: 'POST', body, userId });
  throw new Error('offline_queued');
}

export const workScheduleApi = {
  listWorkSchedules: (userId: string, projectId: string) =>
    req<WorkSchedule[]>(`/api/v1/projects/${projectId}/work-schedules`, {}, userId),

  getActiveWorkSchedule: (
    userId: string,
    projectId: string,
    opts?: { signal?: AbortSignal },
  ) => req<WorkSchedule | null>(
    `/api/v1/projects/${projectId}/work-schedules/active`,
    { signal: opts?.signal, cacheFallback: false },
    userId,
  ),

  /** Absence is accepted only from a successful null response or a real 404. */
  fetchActiveSchedulePlan: async (
    userId: string,
    projectId: string,
    opts?: { signal?: AbortSignal },
  ): Promise<ActiveWorkScheduleResult> => {
    try {
      const data = await req<WorkSchedule | null>(
        `/api/v1/projects/${projectId}/work-schedules/active`,
        { signal: opts?.signal, cacheFallback: false },
        userId,
      );
      return data == null ? { kind: 'absent' } : { kind: 'plan', plan: data };
    } catch (error) {
      if (error instanceof ApiError && error.status === 404) return { kind: 'absent' };
      throw error;
    }
  },

  createWorkSchedule: async (userId: string, projectId: string, body: WorkScheduleCreateInput = {}) => {
    const requestBody = {
      ...body,
      client_request_id: body.client_request_id ?? createClientRequestId('work-schedule'),
    };
    const payload = JSON.stringify(requestBody);
    const path = `/api/v1/projects/${projectId}/work-schedules`;
    try {
      return await req<WorkSchedule>(path, { method: 'POST', body: payload }, userId);
    } catch (error) {
      return queueReplaySafeScheduleCommand<WorkSchedule>(userId, path, payload, error);
    }
  },

  submitWorkSchedule: async (userId: string, projectId: string, scheduleId: string) => {
    const path = `/api/v1/projects/${projectId}/work-schedules/${scheduleId}/submit`;
    const body = '{}';
    try {
      return await req<WorkSchedule>(path, { method: 'POST' }, userId);
    } catch (error) {
      return queueReplaySafeScheduleCommand<WorkSchedule>(userId, path, body, error);
    }
  },

  confirmWorkSchedule: async (userId: string, projectId: string, scheduleId: string) => {
    const path = `/api/v1/projects/${projectId}/work-schedules/${scheduleId}/confirm`;
    const body = '{}';
    try {
      return await req<WorkSchedule>(path, { method: 'POST' }, userId);
    } catch (error) {
      return queueReplaySafeScheduleCommand<WorkSchedule>(userId, path, body, error);
    }
  },

  rejectWorkSchedule: async (userId: string, projectId: string, scheduleId: string, reason?: string) => {
    const path = `/api/v1/projects/${projectId}/work-schedules/${scheduleId}/reject`;
    const body = JSON.stringify({ reason });
    try {
      return await req<WorkSchedule>(path, { method: 'POST', body }, userId);
    } catch (error) {
      return queueReplaySafeScheduleCommand<WorkSchedule>(userId, path, body, error);
    }
  },

  /**
   * Item-status same-state writes had an authority hole and are not yet request-ledgered.
   * Keep them fresh-only until the exact replay contract is qualified end-to-end.
   */
  updateWorkScheduleItemStatus: (
    userId: string,
    projectId: string,
    scheduleId: string,
    itemId: string,
    body: { status: WorkScheduleItemStatus; blocking_reason?: string; progress_percent?: number },
  ) => req<WorkScheduleItem>(
    `/api/v1/projects/${projectId}/work-schedules/${scheduleId}/items/${itemId}/status`,
    { method: 'POST', body: JSON.stringify(body) },
    userId,
  ),
};